amplify/backend/function/getMembersForColony/src/index.js

```diff
@@ -0,0 +1,202 @@
+const REPUTATION_ENDPOINT = 'http://network-contracts:3002';
+
+const SortingMethods = {
+  BY_HIGHEST_REP: 'BY_HIGHEST_REP',
+  BY_LOWEST_REP: 'BY_LOWEST_REP',
+  // @NOTE - this might be useful in the future
+  // BY_MORE_PERMISSIONS,
+  // BY_LESS_PERMISSIONS,
+};
+
+exports.handler = async (event) => {
+  const {
+    colonyAddress,
+    rootHash,
+    domainId = ROOT_DOMAIN_ID,
+    sortingMethod = SortingMethods.BY_HIGHEST_REP,
+  } = event?.arguments?.input || {};
+  const provider = new providers.JsonRpcProvider(RPC_URL);
+
+  const {
+    etherRouterAddress: networkAddress,
+    // eslint-disable-next-line global-require
+  } = require('../../../../mock-data/colonyNetworkArtifacts/etherrouter-address.json');
+  const networkClient = getColonyNetworkClient(Network.Custom, provider, {
+    networkAddress,
+    reputationOracleEndpoint: REPUTATION_ENDPOINT,
+  });
+
+  const colonyClient = await networkClient.getColonyClient(colonyAddress);
+  const { skillId } = await colonyClient.getDomain(domainId);
+  const { addresses: addressesWithReputation } =
+    await colonyClient.getMembersReputation(skillId);
+
+  // Get total reputation for colony
+  let totalColonyReputation;
+  try {
+    totalColonyReputation = await colonyClient.getReputationWithoutProofs(
+      skillId,
+      constants.AddressZero,
+      rootHash,
+    );
+  } catch (error) {
+    // Not throwing anything, as its possible that the domain does not have
+    // any reputation, and we don't want to break the whole query
+  }
+
+  /*
+   * Validate Colony addresses
+   */
+  let checksummedAddress;
+  try {
+    checksummedAddress = utils.getAddress(colonyAddress);
+  } catch (error) {
+    throw new Error(
+      `Colony address "${colonyAddress}" is not valid (after checksum)`,
+    );
+  }
+
+  // get list of watchers for colony
+  let watchers = [];
+  if (domainId === ALL_DOMAIN_ID || domainId === ROOT_DOMAIN_ID) {
+    const { data, errors } = await graphqlRequest(
+      getWatchersInColony,
+      { id: checksummedAddress },
+      GRAPHQL_URI,
+      API_KEY,
+    );
+
+    if (errors || !data) {
+      const [error] = errors;
+      throw new Error(
+        error?.message || 'Could not fetch colony data from DynamoDB',
+      );
+    }
+
+    // Remove members that have reputation (i.e. contributors) to
+    // leave only watchers
+    watchers = data?.getColonyByAddress?.items[0]?.watchers.items.filter(
+      (item) =>
+        addressesWithReputation?.every(
+          (address) => address.toLowerCase() !== item.user.id.toLowerCase(),
+        ),
+    );
+  }
+
+  // get reputation for each address
+  const contributors = await Promise.all(
+    addressesWithReputation.map(async (address) => {
+      const { data } = await graphqlRequest(
+        getUser,
+        {
+          id: utils.getAddress(address),
+        },
+        GRAPHQL_URI,
+        API_KEY,
+      );
```

Comment:
A good improvement here is that if we could fetch users all at once (ie: `getUsers`), using the addresses array in `addressesWithReputation` as a filter.

That way, we'd only have to do this request once, rather than for each user.

Reply:
I've refactored the algorithm slightly, so this call is. no longer needed. All the user data is already fetched when obtaining the watcher (really member) data.
I split this into either watcher or contributor, so no need for another lookup.